EIPS/eip-dtype_storage_extension.md

```diff
@@ -0,0 +1,136 @@
+
+We propose a Solidity CRUD pattern, as described in https://medium.com/robhitchens/solidity-crud-part-1-824ffa69509a, where records can also be retrieved using their index - a monotonically increasing counter.
+
+An stub implementation for the TypeStorageContract would look like:
+
+```
+import './TypeALib.sol';
+
+contract TypeAStorage {
+    using TypeALib for TypeALib.TypeA;
+
+    bytes32[] public typeIndex;
+    mapping(bytes32 => Type) public typeStruct;
+
+    struct Type {
+        TypeALib.TypeA data;
+        uint256 index;
+    }
+
+    event LogNew(bytes32 indexed identifier, uint256 indexed index);
+    event LogUpdate(bytes32 indexed identifier, uint256 indexed index);
+    event LogRemove(bytes32 indexed identifier, uint256 indexed index);
+
+    function insert(TypeALib.TypeA memory data) public returns (bytes32 identifier);
+
+    function insertBytes(bytes memory data) public returns (bytes32 identifier);
+
+    function remove(bytes32 identifier) public returns(uint256 index);
+
+    function update(bytes32 identifier, TypeALib.TypeA memory data) public returns(bytes32 identifier)
+
+    function isStored(bytes32 identifier) public view returns(bool stored);
+
+    function getByHash(bytes32 identifier) public view returns(TypeALib.TypeA memory data);
+
+    function getByIndex(uint256 index) public view returns(TypeALib.TypeA memory data);
+
+    function count() public view returns(uint256 counter);
+}
+```
+
+## Rationale
+
+We are now thinking about a building block as a smart contract with an encapsulated object that contains state changing functions that are only understood from within. This is more akin to Object-Oriented Programming and poses interoperability and scalability issues. Not necessarily for an individual project, but for a global Ethereum OS. This is why we are proposing to separate data from business logic and data structure definitions.
+
+When you have public aggregated data, categorized on each type, anyone can build tools on top of it. This is a radical change from the closed or dispersed data patterns that we find in web2.
+
+We have chosen to define a `TypeRootContract` instead of extending the dType registry with fields for the TypeStorage contract, because this approach enables easier interface updates in the future. It is more extensible.
+
+The storage pattern used for dType itself and all the Type Storage contracts can be the same. This lowers the cost of building, testing and auditing the code.
+
+The `TypeStorageContract` pattern should ensure:
+- type instance addressability by the primary identifier
+- a way to retrieve all records from the contract
+- counting the number of records
+
+
+## Backwards Compatibility
+
+This proposal does not affect extant Ethereum standards or implementations. It uses the present experimental version of ABIEncoderV2.
```

Comment:
'existent'?

Reply:
`extant` exists, but `existent` is more common, so I'm ok with the changes, thank you.